### PR TITLE
Introduce planner_multi_t 

### PR DESCRIPTION
**Pull request by dongahn**
This PR is still experimental: I thought I should post this now as I will be busy with travel for awhile. 

This PR mainly addresses a minimum time resource tree problem within the planner layer. This reverse search tree is designed to satisify a query like "when is the earliest scheduled point at which I can schedule *n* amounts of resources" in `O(lg N)` time. However, this advanced algorithm only works when it deals with a single resource type and doesn't work correctly when it deals with multiple resource types.

In response, this PR limits the planner layer such that that it only manages a single resource type. Instead, it adds `planner_multi_t` layer to handle multiple resources. It also adjusted the rest of the resource code for this change.

Finally a bunch of test cases has also been added.

When I circle back to this, I will do further cleanup and tighten some loosen ends to ensure correctness for a few more corner cases.

### CODE REVIEW

**Comment by dongahn** [comment]
@SteVwonder and I discussed this. We will try to land this with a minimal amount of work so that @SteVwonder can start to integrate the hwloc reader into the new graph-based infrastructure.

**Comment by SteVwonder** [comment]
@dongahn can you rebase onto master?  I can merge afterwards.

Also, you mentioned in the PR description
> When I circle back to this, I will ... tighten some loosen ends to ensure correctness for a few more corner cases.

Do you happen to remember what those corner cases were?  If so, can you document them in an issue so that we do not forget to circle back around to them?

**Comment by dongahn** [comment]
Yes, I will do those. Thanks!

**Comment by dongahn** [comment]
@SteVwonder: OK. I did rebase this to the upstream master. I think the main thing was to change some method names in traverser classes and etc to increase readability and add generally more testing. Both can come later.

f99c21e is so that users can program prune filters more effectively. But this can also come much later. 

Perhaps I can change this commit as "add a placeholder command line option" This way, completing this functionality by me don't have to be in the way of your progress.

**Comment by SteVwonder** [comment]
Sounds like a good idea

File: resource/planner/planner_multi.h [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+ *  \param len          length of resource_counts and resource_types arrays.
+ *  \return             0 if the request can be satisfied; -1 if it cannot
+ *                      be satisfied or an error encountered (errno as follows):
+ *                          EINVAL: invalid argument.
+ *                          ERANGE: resource_counts contain an out-of-range value.
+ *                          ENOTSUP: internal error encountered.
+ */
+int planner_multi_avail_during (planner_multi_t *ctx, int64_t at,
+                                uint64_t duration,
+                                const uint64_t *resource_requests, size_t len);
```
[the remark this answers is not in the document]
Comment (SteVwonder) [reply resource/planner/planner_multi.h]: Actually, on further thought, maybe for the functions that consume the counts, `resource_requests` makes more sense.  And for functions that copy the counts into the array, `resource_counts` makes more sense.

Idk, I don't have super strong feelings either way.  But there is an inconsistency between the function argument and the comments.

**Comment by dongahn** [comment]
@SteVwonder: Thank you for the review. Yes, I will have one more pass on those names and at least have those names consistent.

**Comment by SteVwonder** [comment]
Does this PR fix #319?

**Comment by dongahn** [comment]
Yes.

**Comment by dongahn** [comment]
@SteVwonder: I think I addressed all of your suggestions. Please take a look and merge if this looks good to you. Thanks.
